### PR TITLE
feat: support consensus tracing

### PR DESCRIPTION
**Pull request by homura**
**What type of PR is this?**
feature
**What this PR does / why we need it**:

The consensus algorithm always plays with a distributed system, debugging or optimizing is so hard. If we can record events to describe what happen when consensus state changes, and then replay with a visualization dashboard, the debugging or optimizing would be easier.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: core/consensus/src/trace.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            address: self.address.clone(),
+        }
+    }
+
+    fn now(&self) -> u64 {
+        now()
+    }
+}
+
+pub fn init_tracer(address: String) {
```
Comment (yejiayu) [inline core/consensus/src/trace.rs]: This should return an error.

File: core/consensus/src/trace.rs [hunk 2]
```diff
@@ -33,7 +33,7 @@ impl Trace for MetricTracer {
 }
 
 pub fn init_tracer(address: String) {
-    if let Err(_) = set_boxed_tracer(Box::new(MetricTracer::new(address))) {
+    if set_boxed_tracer(Box::new(MetricTracer::new(address))).is_err() {
```
Comment (KaoImin) [inline core/consensus/src/trace.rs]: Eat the error here?

File: core/consensus/src/trace.rs [hunk 3]
```diff
@@ -32,8 +32,6 @@ impl Trace for MetricTracer {
     }
 }
 
-pub fn init_tracer(address: String) {
-    if set_boxed_tracer(Box::new(MetricTracer::new(address))).is_err() {
-        error!("tracing: tracing init failed");
-    }
+pub fn init_tracer(address: String) -> Result<(), SetTraceError> {
```
Comment (yejiayu) [inline core/consensus/src/trace.rs]: Return ProtocolResult

File: src/main.rs [hunk 4]
```diff
@@ -266,6 +267,7 @@ async fn start(cfg: &Config) -> ProtocolResult<()> {
         consensus_interval: cfg.consensus.interval,
     };
 
+    init_tracer(my_address.as_hex()).unwrap();
```
Comment (KaoImin) [inline src/main.rs]: Use `init().expect(error message)` here.

File: core/consensus/src/trace.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+use log::error;
```
Comment (KaoImin) [inline core/consensus/src/trace.rs]:
```suggestion
use log::{error, trace};
```

File: core/consensus/src/trace.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+impl MetricTracer {
+    fn new(address: String) -> MetricTracer {
+        MetricTracer { address }
+    }
+}
+
+impl Trace for MetricTracer {
+    fn report(&self, point: TracePoint) {
+        match to_string(&point) {
+            Ok(json) => log::trace!(target: "metrics", "{}", json),
```
Comment (KaoImin) [inline core/consensus/src/trace.rs]:
```suggestion
            Ok(json) => trace!(target: "metrics", "{}", json),
```

File: src/main.rs (hunk 4 above)
Comment (homura) [reply src/main.rs]: > Use `init().expect(error message)` here.

returned ProtocolResult now, the `unwrap` is similar to `expect`

File: core/consensus/src/trace.rs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    fn now(&self) -> u64 {
+        now()
+    }
+}
+
+pub fn init_tracer(address: String) -> ProtocolResult<()> {
+    set_boxed_tracer(Box::new(MetricTracer::new(address))).map_err(|_| {
+        ProtocolError::from(ConsensusError::Other("failed to init tracer ".to_string()))
```
Comment (yejiayu) [inline core/consensus/src/trace.rs]: ConsensusError::Other("failed to init tracer ".to_string().into

Comment (homura) [reply core/consensus/src/trace.rs]: map_err is convenient here, why still use `Err into`

Comment (yejiayu) [reply core/consensus/src/trace.rs]: `ProtocolError::from` is unnecessary.

Try replacing it with `ConsensusError::Other("failed to init tracer ".to_string()).into`

Comment (homura) [reply core/consensus/src/trace.rs]: OK, I know, this looks better

**Pull request merged**